{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2bed9acd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: pandas in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (1.4.2)\n",
      "Collecting pandas\n",
      "  Downloading pandas-2.2.2-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (13.1 MB)\n",
      "\u001b[K     |████████████████████████████████| 13.1 MB 3.6 MB/s eta 0:00:01\n",
      "\u001b[?25hRequirement already satisfied: plotly in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (5.6.0)\n",
      "Collecting plotly\n",
      "  Using cached plotly-5.22.0-py3-none-any.whl (16.4 MB)\n",
      "Collecting numpy>=1.22.4\n",
      "  Downloading numpy-2.0.0-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (19.3 MB)\n",
      "\u001b[K     |████████████████████████████████| 19.3 MB 82.0 MB/s eta 0:00:01\n",
      "\u001b[?25hRequirement already satisfied: pytz>=2020.1 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from pandas) (2021.3)\n",
      "Collecting tzdata>=2022.7\n",
      "  Downloading tzdata-2024.1-py2.py3-none-any.whl (345 kB)\n",
      "\u001b[K     |████████████████████████████████| 345 kB 89.3 MB/s eta 0:00:01\n",
      "\u001b[?25hRequirement already satisfied: python-dateutil>=2.8.2 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from pandas) (2.8.2)\n",
      "Requirement already satisfied: tenacity>=6.2.0 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from plotly) (8.0.1)\n",
      "Requirement already satisfied: packaging in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from plotly) (21.3)\n",
      "Requirement already satisfied: six>=1.5 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from python-dateutil>=2.8.2->pandas) (1.16.0)\n",
      "Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from packaging->plotly) (3.0.4)\n",
      "Installing collected packages: tzdata, numpy, plotly, pandas\n",
      "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
      "daal4py 2021.5.0 requires daal==2021.4.0, which is not installed.\n",
      "scipy 1.7.3 requires numpy<1.23.0,>=1.16.5, but you have numpy 2.0.0 which is incompatible.\n",
      "numba 0.55.1 requires numpy<1.22,>=1.18, but you have numpy 2.0.0 which is incompatible.\u001b[0m\n",
      "Successfully installed numpy-2.0.0 pandas-2.2.2 plotly-5.22.0 tzdata-2024.1\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "#pip install --upgrade pandas plotly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "66eb99e7-6706-44b0-aa66-6481c2876f6b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: scikit-learn in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (1.0.2)\n",
      "Collecting scikit-learn\n",
      "  Downloading scikit_learn-1.5.0-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (13.4 MB)\n",
      "\u001b[K     |████████████████████████████████| 13.4 MB 3.8 MB/s eta 0:00:01\n",
      "\u001b[?25hCollecting threadpoolctl>=3.1.0\n",
      "  Downloading threadpoolctl-3.5.0-py3-none-any.whl (18 kB)\n",
      "Requirement already satisfied: numpy>=1.19.5 in ./.local/lib/python3.9/site-packages (from scikit-learn) (1.22.4)\n",
      "Collecting joblib>=1.2.0\n",
      "  Downloading joblib-1.4.2-py3-none-any.whl (301 kB)\n",
      "\u001b[K     |████████████████████████████████| 301 kB 39.0 MB/s eta 0:00:01\n",
      "\u001b[?25hRequirement already satisfied: scipy>=1.6.0 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from scikit-learn) (1.7.3)\n",
      "Installing collected packages: threadpoolctl, joblib, scikit-learn\n",
      "Successfully installed joblib-1.4.2 scikit-learn-1.5.0 threadpoolctl-3.5.0\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "#pip install --upgrade scikit-learn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "68cbf72d-8877-4235-8753-50681e04846b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: numexpr in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (2.8.1)\n",
      "Collecting numexpr\n",
      "  Using cached numexpr-2.10.1-cp39-cp39-linux_x86_64.whl\n",
      "Requirement already satisfied: bottleneck in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (1.3.4)\n",
      "Collecting bottleneck\n",
      "  Using cached Bottleneck-1.4.0-cp39-cp39-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (355 kB)\n",
      "Requirement already satisfied: numpy>=1.23.0 in ./.local/lib/python3.9/site-packages (from numexpr) (2.0.0)\n",
      "Installing collected packages: numexpr, bottleneck\n",
      "Successfully installed bottleneck-1.4.0 numexpr-2.10.1\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install --upgrade --user numexpr bottleneck"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2293ccc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from sklearn.preprocessing import LabelEncoder, StandardScaler\n",
    "from sklearn.cluster import DBSCAN\n",
    "import plotly.express as px\n",
    "import datetime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "69866097",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_df = pd.read_csv('NLP_Dataset1.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2cbbb4ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AssetId</th>\n",
       "      <th>organizationid</th>\n",
       "      <th>organizationcountrycode</th>\n",
       "      <th>locationid</th>\n",
       "      <th>AssetType</th>\n",
       "      <th>AlarmLabel</th>\n",
       "      <th>AlarmMessage</th>\n",
       "      <th>Severity</th>\n",
       "      <th>ActivatedTimestamp</th>\n",
       "      <th>ClearedTimestamp</th>\n",
       "      <th>month</th>\n",
       "      <th>week</th>\n",
       "      <th>ResolutionTimeMinutes</th>\n",
       "      <th>Lemmas_No_Stop_Words</th>\n",
       "      <th>Tokens</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>001e6d91-324b-4b99-84f8-9435148de9ce</td>\n",
       "      <td>5effecae-fdee-494a-bf38-d4cb5d7cc405</td>\n",
       "      <td>US</td>\n",
       "      <td>Default Location</td>\n",
       "      <td>UPS</td>\n",
       "      <td>Battery Near End Of Life Cleared</td>\n",
       "      <td>The battery is no longer near the end of life.</td>\n",
       "      <td>WARNING</td>\n",
       "      <td>2021-01-17T23:32:55Z</td>\n",
       "      <td>2021-03-16T14:37:52Z</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3</td>\n",
       "      <td>82984.950000</td>\n",
       "      <td>['battery', 'near', 'end', 'life', 'clear']</td>\n",
       "      <td>['battery', 'near', 'end', 'of', 'life', 'clea...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>001e6d91-324b-4b99-84f8-9435148de9ce</td>\n",
       "      <td>5effecae-fdee-494a-bf38-d4cb5d7cc405</td>\n",
       "      <td>US</td>\n",
       "      <td>Default Location</td>\n",
       "      <td>UPS</td>\n",
       "      <td>Power Failure Cleared</td>\n",
       "      <td>No longer on battery power.</td>\n",
       "      <td>WARNING</td>\n",
       "      <td>2021-01-08T20:55:56Z</td>\n",
       "      <td>2021-01-08T20:56:19Z</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.383333</td>\n",
       "      <td>['power', 'failure', 'clear']</td>\n",
       "      <td>['power', 'failure', 'cleared']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>004209ba-f692-41b8-af30-5b51996f1abb</td>\n",
       "      <td>a3768a07-4e2d-4075-ab6b-66e3541fc6c4</td>\n",
       "      <td>US</td>\n",
       "      <td>275102b5-af38-456c-982b-7a6d33850e13</td>\n",
       "      <td>UPS</td>\n",
       "      <td>Device Communication Alert</td>\n",
       "      <td>Communication with the UPS \\\"redact\\\" has been...</td>\n",
       "      <td>CRITICAL</td>\n",
       "      <td>2021-01-08T20:28:05Z</td>\n",
       "      <td>2021-01-08T20:36:14Z</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>8.150000</td>\n",
       "      <td>['device', 'communication', 'alert']</td>\n",
       "      <td>['device', 'communication', 'alert']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>004209ba-f692-41b8-af30-5b51996f1abb</td>\n",
       "      <td>a3768a07-4e2d-4075-ab6b-66e3541fc6c4</td>\n",
       "      <td>US</td>\n",
       "      <td>275102b5-af38-456c-982b-7a6d33850e13</td>\n",
       "      <td>UPS</td>\n",
       "      <td>Device Communication Alert</td>\n",
       "      <td>Communication with the UPS \\\"redact\\\" has been...</td>\n",
       "      <td>CRITICAL</td>\n",
       "      <td>2021-01-29T11:51:23Z</td>\n",
       "      <td>2021-01-29T11:56:16Z</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5</td>\n",
       "      <td>4.883333</td>\n",
       "      <td>['device', 'communication', 'alert']</td>\n",
       "      <td>['device', 'communication', 'alert']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>004209ba-f692-41b8-af30-5b51996f1abb</td>\n",
       "      <td>a3768a07-4e2d-4075-ab6b-66e3541fc6c4</td>\n",
       "      <td>US</td>\n",
       "      <td>275102b5-af38-456c-982b-7a6d33850e13</td>\n",
       "      <td>UPS</td>\n",
       "      <td>Device Communication Alert</td>\n",
       "      <td>Communication with the UPS \\\"redact\\\" has been...</td>\n",
       "      <td>CRITICAL</td>\n",
       "      <td>2021-01-29T09:41:23Z</td>\n",
       "      <td>2021-01-29T09:46:22Z</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5</td>\n",
       "      <td>4.983333</td>\n",
       "      <td>['device', 'communication', 'alert']</td>\n",
       "      <td>['device', 'communication', 'alert']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                AssetId                        organizationid  \\\n",
       "0  001e6d91-324b-4b99-84f8-9435148de9ce  5effecae-fdee-494a-bf38-d4cb5d7cc405   \n",
       "1  001e6d91-324b-4b99-84f8-9435148de9ce  5effecae-fdee-494a-bf38-d4cb5d7cc405   \n",
       "2  004209ba-f692-41b8-af30-5b51996f1abb  a3768a07-4e2d-4075-ab6b-66e3541fc6c4   \n",
       "3  004209ba-f692-41b8-af30-5b51996f1abb  a3768a07-4e2d-4075-ab6b-66e3541fc6c4   \n",
       "4  004209ba-f692-41b8-af30-5b51996f1abb  a3768a07-4e2d-4075-ab6b-66e3541fc6c4   \n",
       "\n",
       "  organizationcountrycode                            locationid AssetType  \\\n",
       "0                      US                      Default Location       UPS   \n",
       "1                      US                      Default Location       UPS   \n",
       "2                      US  275102b5-af38-456c-982b-7a6d33850e13       UPS   \n",
       "3                      US  275102b5-af38-456c-982b-7a6d33850e13       UPS   \n",
       "4                      US  275102b5-af38-456c-982b-7a6d33850e13       UPS   \n",
       "\n",
       "                         AlarmLabel  \\\n",
       "0  Battery Near End Of Life Cleared   \n",
       "1             Power Failure Cleared   \n",
       "2        Device Communication Alert   \n",
       "3        Device Communication Alert   \n",
       "4        Device Communication Alert   \n",
       "\n",
       "                                        AlarmMessage  Severity  \\\n",
       "0     The battery is no longer near the end of life.   WARNING   \n",
       "1                        No longer on battery power.   WARNING   \n",
       "2  Communication with the UPS \\\"redact\\\" has been...  CRITICAL   \n",
       "3  Communication with the UPS \\\"redact\\\" has been...  CRITICAL   \n",
       "4  Communication with the UPS \\\"redact\\\" has been...  CRITICAL   \n",
       "\n",
       "     ActivatedTimestamp      ClearedTimestamp  month  week  \\\n",
       "0  2021-01-17T23:32:55Z  2021-03-16T14:37:52Z    1.0     3   \n",
       "1  2021-01-08T20:55:56Z  2021-01-08T20:56:19Z    1.0     2   \n",
       "2  2021-01-08T20:28:05Z  2021-01-08T20:36:14Z    1.0     2   \n",
       "3  2021-01-29T11:51:23Z  2021-01-29T11:56:16Z    1.0     5   \n",
       "4  2021-01-29T09:41:23Z  2021-01-29T09:46:22Z    1.0     5   \n",
       "\n",
       "   ResolutionTimeMinutes                         Lemmas_No_Stop_Words  \\\n",
       "0           82984.950000  ['battery', 'near', 'end', 'life', 'clear']   \n",
       "1               0.383333                ['power', 'failure', 'clear']   \n",
       "2               8.150000         ['device', 'communication', 'alert']   \n",
       "3               4.883333         ['device', 'communication', 'alert']   \n",
       "4               4.983333         ['device', 'communication', 'alert']   \n",
       "\n",
       "                                              Tokens  \n",
       "0  ['battery', 'near', 'end', 'of', 'life', 'clea...  \n",
       "1                    ['power', 'failure', 'cleared']  \n",
       "2               ['device', 'communication', 'alert']  \n",
       "3               ['device', 'communication', 'alert']  \n",
       "4               ['device', 'communication', 'alert']  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c2aa27a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  organizationcountrycode  Severity               AssetType  count\n",
      "0                      AE  CRITICAL                    CRAC     58\n",
      "1                      AE  CRITICAL                     EMS   1334\n",
      "2                      AE  CRITICAL  NETBOTZ_RACK_APPLIANCE     33\n",
      "3                      AE  CRITICAL                     POD    309\n",
      "4                      AE  CRITICAL                    RPDU  56363\n"
     ]
    }
   ],
   "source": [
    "grouped_df = sample_df.groupby(['organizationcountrycode', 'Severity','AssetType']).size().reset_index(name='count')\n",
    "\n",
    "# Display the grouped DataFrame\n",
    "print(grouped_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a53d956c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "organizationcountrycode    object\n",
       "Severity                   object\n",
       "AssetType                  object\n",
       "count                       int64\n",
       "dtype: object"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grouped_df.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7e1d3c1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ensure correct data types\n",
    "grouped_df['organizationcountrycode'] = grouped_df['organizationcountrycode'].astype(str)\n",
    "grouped_df['Severity'] = grouped_df['Severity'].astype(str)\n",
    "grouped_df['AssetType'] = grouped_df['AssetType'].astype(str)\n",
    "grouped_df['count'] = grouped_df['count'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f5ad5293",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Collecting jupyter_dash\n",
      "  Downloading jupyter_dash-0.4.2-py3-none-any.whl (23 kB)\n",
      "Requirement already satisfied: nest-asyncio in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from jupyter_dash) (1.5.5)\n",
      "Requirement already satisfied: ipython in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from jupyter_dash) (8.2.0)\n",
      "Collecting ansi2html\n",
      "  Downloading ansi2html-1.9.2-py3-none-any.whl (17 kB)\n",
      "Requirement already satisfied: flask in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from jupyter_dash) (1.1.2)\n",
      "Requirement already satisfied: ipykernel in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from jupyter_dash) (6.9.1)\n",
      "Requirement already satisfied: retrying in ./.local/lib/python3.9/site-packages (from jupyter_dash) (1.3.4)\n",
      "Requirement already satisfied: requests in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from jupyter_dash) (2.27.1)\n",
      "Requirement already satisfied: dash in ./.local/lib/python3.9/site-packages (from jupyter_dash) (2.17.1)\n",
      "Requirement already satisfied: plotly>=5.0.0 in ./.local/lib/python3.9/site-packages (from dash->jupyter_dash) (5.22.0)\n",
      "Requirement already satisfied: dash-table==5.0.0 in ./.local/lib/python3.9/site-packages (from dash->jupyter_dash) (5.0.0)\n",
      "Requirement already satisfied: dash-html-components==2.0.0 in ./.local/lib/python3.9/site-packages (from dash->jupyter_dash) (2.0.0)\n",
      "Requirement already satisfied: importlib-metadata in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from dash->jupyter_dash) (4.11.3)\n",
      "Requirement already satisfied: setuptools in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from dash->jupyter_dash) (61.2.0)\n",
      "Requirement already satisfied: typing-extensions>=4.1.1 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from dash->jupyter_dash) (4.1.1)\n",
      "Requirement already satisfied: Werkzeug<3.1 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from dash->jupyter_dash) (2.0.3)\n",
      "Requirement already satisfied: dash-core-components==2.0.0 in ./.local/lib/python3.9/site-packages (from dash->jupyter_dash) (2.0.0)\n",
      "Requirement already satisfied: click>=5.1 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from flask->jupyter_dash) (8.0.4)\n",
      "Requirement already satisfied: itsdangerous>=0.24 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from flask->jupyter_dash) (2.0.1)\n",
      "Requirement already satisfied: Jinja2>=2.10.1 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from flask->jupyter_dash) (2.11.3)\n",
      "Requirement already satisfied: MarkupSafe>=0.23 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from Jinja2>=2.10.1->flask->jupyter_dash) (2.0.1)\n",
      "Requirement already satisfied: tenacity>=6.2.0 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from plotly>=5.0.0->dash->jupyter_dash) (8.0.1)\n",
      "Requirement already satisfied: packaging in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from plotly>=5.0.0->dash->jupyter_dash) (21.3)\n",
      "Requirement already satisfied: zipp>=0.5 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from importlib-metadata->dash->jupyter_dash) (3.7.0)\n",
      "Requirement already satisfied: tornado<7.0,>=4.2 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from ipykernel->jupyter_dash) (6.1)\n",
      "Requirement already satisfied: jupyter-client<8.0 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from ipykernel->jupyter_dash) (6.1.12)\n",
      "Requirement already satisfied: matplotlib-inline<0.2.0,>=0.1.0 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from ipykernel->jupyter_dash) (0.1.2)\n",
      "Requirement already satisfied: debugpy<2.0,>=1.0.0 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from ipykernel->jupyter_dash) (1.5.1)\n",
      "Requirement already satisfied: traitlets<6.0,>=5.1.0 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from ipykernel->jupyter_dash) (5.1.1)\n",
      "Requirement already satisfied: pygments>=2.4.0 in ./.local/lib/python3.9/site-packages (from ipython->jupyter_dash) (2.18.0)\n",
      "Requirement already satisfied: pexpect>4.3 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from ipython->jupyter_dash) (4.8.0)\n",
      "Requirement already satisfied: decorator in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from ipython->jupyter_dash) (5.1.1)\n",
      "Requirement already satisfied: pickleshare in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from ipython->jupyter_dash) (0.7.5)\n",
      "Requirement already satisfied: prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from ipython->jupyter_dash) (3.0.20)\n",
      "Requirement already satisfied: stack-data in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from ipython->jupyter_dash) (0.2.0)\n",
      "Requirement already satisfied: backcall in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from ipython->jupyter_dash) (0.2.0)\n",
      "Requirement already satisfied: jedi>=0.16 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from ipython->jupyter_dash) (0.18.1)\n",
      "Requirement already satisfied: parso<0.9.0,>=0.8.0 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from jedi>=0.16->ipython->jupyter_dash) (0.8.3)\n",
      "Requirement already satisfied: pyzmq>=13 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from jupyter-client<8.0->ipykernel->jupyter_dash) (22.3.0)\n",
      "Requirement already satisfied: jupyter-core>=4.6.0 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from jupyter-client<8.0->ipykernel->jupyter_dash) (4.9.2)\n",
      "Requirement already satisfied: python-dateutil>=2.1 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from jupyter-client<8.0->ipykernel->jupyter_dash) (2.8.2)\n",
      "Requirement already satisfied: ptyprocess>=0.5 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from pexpect>4.3->ipython->jupyter_dash) (0.7.0)\n",
      "Requirement already satisfied: wcwidth in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0->ipython->jupyter_dash) (0.2.5)\n",
      "Requirement already satisfied: six>=1.5 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from python-dateutil>=2.1->jupyter-client<8.0->ipykernel->jupyter_dash) (1.16.0)\n",
      "Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from packaging->plotly>=5.0.0->dash->jupyter_dash) (3.0.4)\n",
      "Requirement already satisfied: charset-normalizer~=2.0.0 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from requests->jupyter_dash) (2.0.4)\n",
      "Requirement already satisfied: idna<4,>=2.5 in ./.local/lib/python3.9/site-packages (from requests->jupyter_dash) (2.10)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from requests->jupyter_dash) (2021.10.8)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from requests->jupyter_dash) (1.26.9)\n",
      "Requirement already satisfied: pure-eval in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from stack-data->ipython->jupyter_dash) (0.2.2)\n",
      "Requirement already satisfied: asttokens in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from stack-data->ipython->jupyter_dash) (2.0.5)\n",
      "Requirement already satisfied: executing in /shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages (from stack-data->ipython->jupyter_dash) (0.8.3)\n",
      "Installing collected packages: ansi2html, jupyter-dash\n",
      "Successfully installed ansi2html-1.9.2 jupyter-dash-0.4.2\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install jupyter_dash"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "5de98e28",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<!DOCTYPE html PUBLIC \"-//W3C//DTD HTML 4.01//EN\" \"http://www.w3.org/TR/html4/strict.dtd\">\n",
       "<html><head>\n",
       "<meta type=\"copyright\" content=\"Copyright (C) 1996-2016 The Squid Software Foundation and contributors\">\n",
       "<meta http-equiv=\"Content-Type\" content=\"text/html; charset=utf-8\">\n",
       "<title>ERROR: The requested URL could not be retrieved</title>\n",
       "<style type=\"text/css\"><!--\n",
       " /*\n",
       " * Copyright (C) 1996-2016 The Squid Software Foundation and contributors\n",
       " *\n",
       " * Squid software is distributed under GPLv2+ license and includes\n",
       " * contributions from numerous individuals and organizations.\n",
       " * Please see the COPYING and CONTRIBUTORS files for details.\n",
       " */\n",
       "\n",
       "/*\n",
       " Stylesheet for Squid Error pages\n",
       " Adapted from design by Free CSS Templates\n",
       " http://www.freecsstemplates.org\n",
       " Released for free under a Creative Commons Attribution 2.5 License\n",
       "*/\n",
       "\n",
       "/* Page basics */\n",
       "* {\n",
       "\tfont-family: verdana, sans-serif;\n",
       "}\n",
       "\n",
       "html body {\n",
       "\tmargin: 0;\n",
       "\tpadding: 0;\n",
       "\tbackground: #efefef;\n",
       "\tfont-size: 12px;\n",
       "\tcolor: #1e1e1e;\n",
       "}\n",
       "\n",
       "/* Page displayed title area */\n",
       "#titles {\n",
       "\tmargin-left: 15px;\n",
       "\tpadding: 10px;\n",
       "\tpadding-left: 100px;\n",
       "\tbackground: url('/squid-internal-static/icons/SN.png') no-repeat left;\n",
       "}\n",
       "\n",
       "/* initial title */\n",
       "#titles h1 {\n",
       "\tcolor: #000000;\n",
       "}\n",
       "#titles h2 {\n",
       "\tcolor: #000000;\n",
       "}\n",
       "\n",
       "/* special event: FTP success page titles */\n",
       "#titles ftpsuccess {\n",
       "\tbackground-color:#00ff00;\n",
       "\twidth:100%;\n",
       "}\n",
       "\n",
       "/* Page displayed body content area */\n",
       "#content {\n",
       "\tpadding: 10px;\n",
       "\tbackground: #ffffff;\n",
       "}\n",
       "\n",
       "/* General text */\n",
       "p {\n",
       "}\n",
       "\n",
       "/* error brief description */\n",
       "#error p {\n",
       "}\n",
       "\n",
       "/* some data which may have caused the problem */\n",
       "#data {\n",
       "}\n",
       "\n",
       "/* the error message received from the system or other software */\n",
       "#sysmsg {\n",
       "}\n",
       "\n",
       "pre {\n",
       "    font-family:sans-serif;\n",
       "}\n",
       "\n",
       "/* special event: FTP / Gopher directory listing */\n",
       "#dirmsg {\n",
       "    font-family: courier;\n",
       "    color: black;\n",
       "    font-size: 10pt;\n",
       "}\n",
       "#dirlisting {\n",
       "    margin-left: 2%;\n",
       "    margin-right: 2%;\n",
       "}\n",
       "#dirlisting tr.entry td.icon,td.filename,td.size,td.date {\n",
       "    border-bottom: groove;\n",
       "}\n",
       "#dirlisting td.size {\n",
       "    width: 50px;\n",
       "    text-align: right;\n",
       "    padding-right: 5px;\n",
       "}\n",
       "\n",
       "/* horizontal lines */\n",
       "hr {\n",
       "\tmargin: 0;\n",
       "}\n",
       "\n",
       "/* page displayed footer area */\n",
       "#footer {\n",
       "\tfont-size: 9px;\n",
       "\tpadding-left: 10px;\n",
       "}\n",
       "\n",
       "\n",
       "body\n",
       ":lang(fa) { direction: rtl; font-size: 100%; font-family: Tahoma, Roya, sans-serif; float: right; }\n",
       ":lang(he) { direction: rtl; }\n",
       " --></style>\n",
       "</head><body id=ERR_ACCESS_DENIED>\n",
       "<div id=\"titles\">\n",
       "<h1>ERROR</h1>\n",
       "<h2>The requested URL could not be retrieved</h2>\n",
       "</div>\n",
       "<hr>\n",
       "\n",
       "<div id=\"content\">\n",
       "<p>The following error was encountered while trying to retrieve the URL: <a href=\"http://127.0.0.1:36732/_alive_9255b7a8-1b70-4378-86b1-ccebc8aaf554\">http://127.0.0.1:36732/_alive_9255b7a8-1b70-4378-86b1-ccebc8aaf554</a></p>\n",
       "\n",
       "<blockquote id=\"error\">\n",
       "<p><b>Access Denied.</b></p>\n",
       "</blockquote>\n",
       "\n",
       "<p>Access control configuration prevents your request from being allowed at this time. Please contact your service provider if you feel this is incorrect.</p>\n",
       "\n",
       "<p>Your cache administrator is <a href=\"mailto:root?subject=CacheErrorInfo%20-%20ERR_ACCESS_DENIED&amp;body=CacheHost%3A%20xfer-00%0D%0AErrPage%3A%20ERR_ACCESS_DENIED%0D%0AErr%3A%20%5Bnone%5D%0D%0ATimeStamp%3A%20Sun,%2023%20Jun%202024%2002%3A04%3A09%20GMT%0D%0A%0D%0AClientIP%3A%2010.99.250.115%0D%0A%0D%0AHTTP%20Request%3A%0D%0AGET%20%2F_alive_9255b7a8-1b70-4378-86b1-ccebc8aaf554%20HTTP%2F1.1%0AUser-Agent%3A%20python-requests%2F2.27.1%0D%0AAccept-Encoding%3A%20gzip,%20deflate,%20br%0D%0AAccept%3A%20*%2F*%0D%0AConnection%3A%20keep-alive%0D%0AHost%3A%20127.0.0.1%3A36732%0D%0A%0D%0A%0D%0A\">root</a>.</p>\n",
       "<br>\n",
       "</div>\n",
       "\n",
       "<hr>\n",
       "<div id=\"footer\">\n",
       "<p>Generated Sun, 23 Jun 2024 02:04:09 GMT by xfer-00 (squid/3.5.20)</p>\n",
       "<!-- ERR_ACCESS_DENIED -->\n",
       "</div>\n",
       "</body></html>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "ename": "OSError",
     "evalue": "Address 'http://127.0.0.1:36732' already in use.\n    Try passing a different port to run_server.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mOSError\u001b[0m                                   Traceback (most recent call last)",
      "Input \u001b[0;32mIn [42]\u001b[0m, in \u001b[0;36m<cell line: 71>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     68\u001b[0m available_port \u001b[38;5;241m=\u001b[39m find_free_port()\n\u001b[1;32m     69\u001b[0m \u001b[38;5;66;03m#print(f\"Available port: {available_port}\")\u001b[39;00m\n\u001b[1;32m     70\u001b[0m \u001b[38;5;66;03m# Run the app\u001b[39;00m\n\u001b[0;32m---> 71\u001b[0m \u001b[43mapp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun_server\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmode\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43minline\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mport\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfind_free_port\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/jupyter_dash/jupyter_app.py:339\u001b[0m, in \u001b[0;36mJupyterDash.run_server\u001b[0;34m(self, mode, width, height, inline_exceptions, **kwargs)\u001b[0m\n\u001b[1;32m    329\u001b[0m         url \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhttp://\u001b[39m\u001b[38;5;132;01m{host}\u001b[39;00m\u001b[38;5;124m:\u001b[39m\u001b[38;5;132;01m{port}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mformat(\n\u001b[1;32m    330\u001b[0m             host\u001b[38;5;241m=\u001b[39mhost, port\u001b[38;5;241m=\u001b[39mport, token\u001b[38;5;241m=\u001b[39mJupyterDash\u001b[38;5;241m.\u001b[39m_token\n\u001b[1;32m    331\u001b[0m         )\n\u001b[1;32m    332\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mOSError\u001b[39;00m(\n\u001b[1;32m    333\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mAddress \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{url}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m already in use.\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    334\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m    Try passing a different port to run_server.\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mformat(\n\u001b[1;32m    335\u001b[0m                 url\u001b[38;5;241m=\u001b[39murl\n\u001b[1;32m    336\u001b[0m             )\n\u001b[1;32m    337\u001b[0m         )\n\u001b[0;32m--> 339\u001b[0m \u001b[43mwait_for_app\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    341\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m JupyterDash\u001b[38;5;241m.\u001b[39m_in_colab:\n\u001b[1;32m    342\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_display_in_colab(dashboard_url, port, mode, width, height)\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/retrying.py:56\u001b[0m, in \u001b[0;36mretry.<locals>.wrap.<locals>.wrapped_f\u001b[0;34m(*args, **kw)\u001b[0m\n\u001b[1;32m     54\u001b[0m \u001b[38;5;129m@six\u001b[39m\u001b[38;5;241m.\u001b[39mwraps(f)\n\u001b[1;32m     55\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mwrapped_f\u001b[39m(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkw):\n\u001b[0;32m---> 56\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mRetrying\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mdargs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mdkw\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcall\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkw\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/retrying.py:266\u001b[0m, in \u001b[0;36mRetrying.call\u001b[0;34m(self, fn, *args, **kwargs)\u001b[0m\n\u001b[1;32m    263\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstop(attempt_number, delay_since_first_attempt_ms):\n\u001b[1;32m    264\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_wrap_exception \u001b[38;5;129;01mand\u001b[39;00m attempt\u001b[38;5;241m.\u001b[39mhas_exception:\n\u001b[1;32m    265\u001b[0m         \u001b[38;5;66;03m# get() on an attempt with an exception should cause it to be raised, but raise just in case\u001b[39;00m\n\u001b[0;32m--> 266\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[43mattempt\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    267\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    268\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m RetryError(attempt)\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/retrying.py:301\u001b[0m, in \u001b[0;36mAttempt.get\u001b[0;34m(self, wrap_exception)\u001b[0m\n\u001b[1;32m    299\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m RetryError(\u001b[38;5;28mself\u001b[39m)\n\u001b[1;32m    300\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 301\u001b[0m         \u001b[43msix\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mreraise\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvalue\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvalue\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvalue\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m2\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    302\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    303\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mvalue\n",
      "File \u001b[0;32m/shared/centos7/anaconda3/2022.05/lib/python3.9/site-packages/six.py:719\u001b[0m, in \u001b[0;36mreraise\u001b[0;34m(tp, value, tb)\u001b[0m\n\u001b[1;32m    717\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m value\u001b[38;5;241m.\u001b[39m__traceback__ \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m tb:\n\u001b[1;32m    718\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m value\u001b[38;5;241m.\u001b[39mwith_traceback(tb)\n\u001b[0;32m--> 719\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m value\n\u001b[1;32m    720\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m    721\u001b[0m     value \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/retrying.py:251\u001b[0m, in \u001b[0;36mRetrying.call\u001b[0;34m(self, fn, *args, **kwargs)\u001b[0m\n\u001b[1;32m    248\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_before_attempts(attempt_number)\n\u001b[1;32m    250\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 251\u001b[0m     attempt \u001b[38;5;241m=\u001b[39m Attempt(\u001b[43mfn\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m, attempt_number, \u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[1;32m    252\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m:\n\u001b[1;32m    253\u001b[0m     tb \u001b[38;5;241m=\u001b[39m sys\u001b[38;5;241m.\u001b[39mexc_info()\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/jupyter_dash/jupyter_app.py:332\u001b[0m, in \u001b[0;36mJupyterDash.run_server.<locals>.wait_for_app\u001b[0;34m()\u001b[0m\n\u001b[1;32m    328\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m res \u001b[38;5;241m!=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mAlive\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m    329\u001b[0m     url \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhttp://\u001b[39m\u001b[38;5;132;01m{host}\u001b[39;00m\u001b[38;5;124m:\u001b[39m\u001b[38;5;132;01m{port}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mformat(\n\u001b[1;32m    330\u001b[0m         host\u001b[38;5;241m=\u001b[39mhost, port\u001b[38;5;241m=\u001b[39mport, token\u001b[38;5;241m=\u001b[39mJupyterDash\u001b[38;5;241m.\u001b[39m_token\n\u001b[1;32m    331\u001b[0m     )\n\u001b[0;32m--> 332\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mOSError\u001b[39;00m(\n\u001b[1;32m    333\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mAddress \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{url}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m already in use.\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    334\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m    Try passing a different port to run_server.\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mformat(\n\u001b[1;32m    335\u001b[0m             url\u001b[38;5;241m=\u001b[39murl\n\u001b[1;32m    336\u001b[0m         )\n\u001b[1;32m    337\u001b[0m     )\n",
      "\u001b[0;31mOSError\u001b[0m: Address 'http://127.0.0.1:36732' already in use.\n    Try passing a different port to run_server."
     ]
    }
   ],
   "source": [
    "from jupyter_dash import JupyterDash\n",
    "from dash import dcc, html\n",
    "from dash.dependencies import Input, Output\n",
    "import plotly.express as px\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "# Create the Dash app\n",
    "app = JupyterDash(__name__)\n",
    "\n",
    "# Layout of the app\n",
    "app.layout = html.Div([\n",
    "    html.H1(\"Treemap of Alarms Clustered by Country, Severity, and Asset Type\"),\n",
    "    \n",
    "    html.Label(\"Select Country:\"),\n",
    "    dcc.Dropdown(\n",
    "        id='country-dropdown',\n",
    "        options=[{'label': country, 'value': country} for country in grouped_df['organizationcountrycode'].unique()],\n",
    "        multi=True,\n",
    "        value=list(grouped_df['organizationcountrycode'].unique())\n",
    "    ),\n",
    "    \n",
    "    html.Label(\"Select Severity:\"),\n",
    "    dcc.Dropdown(\n",
    "        id='severity-dropdown',\n",
    "        options=[{'label': severity, 'value': severity} for severity in grouped_df['Severity'].unique()],\n",
    "        multi=True,\n",
    "        value=list(grouped_df['Severity'].unique())\n",
    "    ),\n",
    "    \n",
    "    dcc.Graph(id='treemap-plot')\n",
    "])\n",
    "\n",
    "# Callback to update the treemap plot based on selected filters\n",
    "@app.callback(\n",
    "    Output('treemap-plot', 'figure'),\n",
    "    [Input('country-dropdown', 'value'),\n",
    "     Input('severity-dropdown', 'value')]\n",
    ")\n",
    "def update_treemap(selected_countries, selected_severities):\n",
    "    filtered_df = grouped_df[(grouped_df['organizationcountrycode'].isin(selected_countries)) & \n",
    "                             (grouped_df['Severity'].isin(selected_severities))]\n",
    "    \n",
    "    fig = px.treemap(\n",
    "        filtered_df,\n",
    "        path=[px.Constant('World'),'organizationcountrycode', 'Severity', 'AssetType'],\n",
    "        values='count',  # This will count the number of rows per category\n",
    "        title='Treemap of Alarms Clustered by Country, Severity, and Asset Type',\n",
    "        labels =['count'],\n",
    "        hover_data=['count']  # Include the count in hover data\n",
    "    )\n",
    "\n",
    "    fig.update_layout(\n",
    "        treemapcolorway=['green','#FF0000', '#FF8C00', '#FFD700', '#ADFF2F', '#00FF00']\n",
    "    )\n",
    "\n",
    "    return fig\n",
    "import socket\n",
    "\n",
    "def find_free_port():\n",
    "    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)\n",
    "    s.bind(('', 0))  # Bind to a free port provided by the host\n",
    "    port = s.getsockname()[1]\n",
    "    s.close()\n",
    "    return port\n",
    "\n",
    "# Find an available port\n",
    "available_port = find_free_port()\n",
    "#print(f\"Available port: {available_port}\")\n",
    "# Run the app\n",
    "app.run_server(mode='inline', port=find_free_port())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "dc11b6ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Available port: 45336\n",
      "Available port: 40593\n",
      "Available port: 52207\n",
      "Available port: 37513\n",
      "Available port: 42263\n",
      "Available port: 54164\n",
      "Available port: 60725\n",
      "Available port: 50662\n",
      "Available port: 44012\n"
     ]
    }
   ],
   "source": [
    "import socket\n",
    "\n",
    "def find_free_port():\n",
    "    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)\n",
    "    s.bind(('', 0))  # Bind to a free port provided by the host\n",
    "    port = s.getsockname()[1]\n",
    "    s.close()\n",
    "    return port\n",
    "\n",
    "# Find an available port\n",
    "for i in range (1,10):\n",
    "    available_port = find_free_port()\n",
    "    print(f\"Available port: {available_port}\")\n",
    "    52124,59507,59507"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "429abdf1-89aa-4ad7-a9a0-c930a78ff55c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fde2e99-c1af-460e-b485-5edb9fb6c1e7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
